src/wallet/service.go

```diff
@@ -0,0 +1,231 @@
+package wallet
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/skycoin/skycoin/src/cipher"
+	"github.com/skycoin/skycoin/src/coin"
+	"github.com/skycoin/skycoin/src/visor/blockdb"
+)
+
+// Service wallet service struct
+type Service struct {
+	sync.RWMutex
+	wallets        Wallets
+	options        []Option
+	firstAddrIDMap map[string]string // key: first address in wallet, value: wallet id
+
+	WalletDirectory string
+}
+
+// NewService new wallet service
+func NewService(walletDir string, options ...Option) (*Service, error) {
+	serv := &Service{
+		firstAddrIDMap: make(map[string]string),
+	}
+	if err := os.MkdirAll(walletDir, os.FileMode(0700)); err != nil {
+		return nil, fmt.Errorf("failed to create wallet directory %s: %v", walletDir, err)
+	}
+
+	serv.WalletDirectory = walletDir
+	for i := range options {
+		serv.options = append(serv.options, options[i])
+	}
+
+	w, err := LoadWallets(serv.WalletDirectory)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load all wallets: %v", err)
+	}
+
+	serv.wallets = serv.removeDup(w)
+
+	if len(serv.wallets) == 0 {
+		wltName := NewWalletFilename()
+		// create default wallet
+		w, err := serv.CreateWallet(wltName)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := w.Save(serv.WalletDirectory); err != nil {
+			return nil, fmt.Errorf("failed to save wallets to %s: %v", serv.WalletDirectory, err)
+		}
+	}
+
+	return serv, nil
+}
+
+// CreateWallet creates wallet
+func (serv *Service) CreateWallet(wltName string, options ...Option) (Wallet, error) {
+	ops := make([]Option, 0, len(serv.options)+len(options))
+	ops = append(ops, serv.options...)
+	ops = append(ops, options...)
+	w, err := NewWallet(wltName, ops...)
+	if err != nil {
+		return Wallet{}, err
+	}
+
+	// generate a default address
+	w.GenerateAddresses(1)
+
+	serv.Lock()
+	defer serv.Unlock()
+	// check dup
+	if id, ok := serv.firstAddrIDMap[w.Entries[0].Address.String()]; ok {
+		return Wallet{}, fmt.Errorf("duplicate wallet with %v", id)
+	}
+
+	if err := serv.wallets.Add(*w); err != nil {
+		return Wallet{}, err
+	}
+
+	if err := w.Save(serv.WalletDirectory); err != nil {
```

Comment:
Remove the added wallet from serv.wallets

Reply:
Why remove the added wallet? if the wallet name is dup, this function should return error, the caller will rename and try again. If the first address is duplicate, we choose the old wallet, cause the wallet entry number in old wallet is >=1, but the new created wallet is 1.

Reply:
oh i thought it would be disk error, then we would show a loaded wallet but it wouldnt be saved, so user would lose it mistakenly

Reply:
Oh, yes, I've updated the code